### PR TITLE
[SYCL][SCLA] Add `aligned_private_alloca` support

### PR DESCRIPTION
**Pull request by victor-eds**
Add support for `aligned_private_alloca`. Use new `__builtin_intel_sycl_alloca_with_align` builtin in implementation.

Revamp diagnostic messages to support the new builtin.

Add SPIR-V integration tests.

### CODE REVIEW

**Review by maarquitos14 — COMMENTED** [review]
llvm-runtime-reviewers changes look good to me. Just one question: should we also test invalid usage? There's a `ValidUsage` directory under `tests-e2e/PrivateAlloca`, should we also have `InvalidUsage` testing negative cases?

**Comment by victor-eds** [comment]
> llvm-runtime-reviewers changes look good to me. Just one question: should we also test invalid usage? There's a `ValidUsage` directory under `tests-e2e/PrivateAlloca`, should we also have `InvalidUsage` testing negative cases?

We have `UnsupportedDevices` to cover usage of the extension when no device supports it. Sema tests covers all other invalid usages of the extension.

**Comment by victor-eds** [comment]
Unrelated CI failures

**Comment by victor-eds** [comment]
@intel/dpcpp-cfe-reviewers can I get a review here?

File: clang/lib/Sema/SemaChecking.cpp [hunk 1]
```diff
@@ -7825,10 +7841,11 @@ bool Sema::CheckIntelSYCLAllocaBuiltinFunctionCall(unsigned, CallExpr *Call) {
     return true;
 
   // Check three template arguments are passed
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaChecking.cpp:7843]: Please update this comment

File: clang/lib/Sema/SemaChecking.cpp [hunk 2]
```diff
@@ -7883,23 +7904,43 @@ bool Sema::CheckIntelSYCLAllocaBuiltinFunctionCall(unsigned, CallExpr *Call) {
...
     else
-      Diag(Loc, diag::err_intel_sycl_alloca_wrong_size) << TA << Ty;
+      D << Ty;
     return true;
   }
 
+  if (IsAlignedAlloca) {
+    TemplateArgument AlignmentArg = CST->get(AlignmentIndex);
+    CharUnits RequestedAlign =
+        CharUnits::fromQuantity(AlignmentArg.getAsIntegral().getZExtValue());
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaChecking.cpp:7929]: I think you can just use `getAsIntegral`  and do these checks. See `SemaBuiltinAllocaWithAlign `

File: clang/test/SemaSYCL/Inputs/private_alloca.hpp [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 #include "./sycl.hpp"
 
+#include <stddef.h>
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/Inputs/private_alloca.hpp:5]: I'm not sure about adding stddef to lit tests. @premanandrao @AaronBallman can you comment here?

File: clang/lib/Sema/SemaChecking.cpp [hunk 4]
```diff
@@ -7883,23 +7904,43 @@ bool Sema::CheckIntelSYCLAllocaBuiltinFunctionCall(unsigned, CallExpr *Call) {
...
-          << DefaultSize.getSExtValue();
+          << IsAlignedAlloca << DefaultSize.getSExtValue();
     return false;
   };
-  const TemplateArgumentList *CST = FD->getTemplateSpecializationArgs();
   if (CheckSize(getASTContext(), Loc, CST)) {
-    TemplateArgument TA = CST->get(1);
+    TemplateArgument TA = CST->get(SpecNameIndex);
     QualType Ty = TA.getNonTypeTemplateArgumentType();
+    const SemaDiagnosticBuilder &D =
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply clang/lib/Sema/SemaChecking.cpp:7916]: Cleaner code as different values are to be printed in both branches of the if-then-else statement below

File: clang/test/SemaSYCL/Inputs/private_alloca.hpp (hunk 3 above)
Comment (victor-eds) [reply clang/test/SemaSYCL/Inputs/private_alloca.hpp:5]: Other tests in `SemaCXX` are doing this. I don't think this is an issue

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 5]
```diff
@@ -23889,17 +23890,30 @@ RValue CodeGenFunction::EmitIntelSYCLPtrAnnotationBuiltin(const CallExpr *E) {
...
+          BuiltinID == Builtin::BI__builtin_intel_sycl_alloca_with_align) &&
+         "Unexpected builtin");
+
+  bool IsAlignedAlloca =
+      BuiltinID == Builtin::BI__builtin_intel_sycl_alloca_with_align;
+
+  constexpr unsigned InvalidIndex = -1;
+  constexpr unsigned ElementTypeIndex = 0;
+  const unsigned AlignmentIndex = IsAlignedAlloca ? 1 : InvalidIndex;
+  const unsigned SpecNameIndex = IsAlignedAlloca ? 2 : 1;
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply clang/lib/CodeGen/CGBuiltin.cpp:23905]: In the future, we hope the SYCL community will reach an agreement w.r.t. the default value for `sycl::access::decorate` across the SYCL API. We will apply that here. That's why we cannot add arguments to the end.

Order was chosen to mimic functions in C++/SYCL such as `malloc_aligned`

[the remark this answers is not in the document]
Comment (victor-eds) [reply clang/lib/CodeGen/CGBuiltin.cpp:23905]: Original discussion: https://github.com/intel/llvm/pull/12660#discussion_r1498501984

File: clang/test/SemaSYCL/Inputs/private_alloca.hpp (hunk 3 above)
Comment (AaronBallman) [reply clang/test/SemaSYCL/Inputs/private_alloca.hpp:5]: `stddef.h` is generally fine to use in lit tests because it's provided by the compiler and it does not do an `include_next` to something on the system. (Basically, we want lit tests to be hermetic and not rely on anything about the test system, so in general it's not okay to include system headers outside of passing `-ffreestanding` and using the compiler ones, but this case is a bit special.)

Comment (victor-eds) [reply clang/test/SemaSYCL/Inputs/private_alloca.hpp:5]: Cool. Thanks! In that case, we can use this header in particular.

**Review by elizabethandrews — APPROVED** [review]

**Comment by victor-eds** [comment]
@intel/llvm-gatekeepers Jenkins CI failure unrelated to these changes. Can we get this merged?
